File: hw5/hw5_monte.py
import numpy as np

import numpy as np
import random

def exp(x):
	return np.e ** x

if __name__ == '__main__':
	file_name = "hw5_input.txt"
	inputs = list(open(file_name))[0]
	inputs = [float(x) for x in inputs.split()]

	S_t, K, r, q, sigma, T_minus_t, M, n, S_ave_t, passing_time, simulations, repetitions = inputs
	M, n, simulations, repetitions = int(M), int(n), int(simulations), int(repetitions)

	print("Monte Carlo simulation to arithmetic average call.")
	print("S_t = {}, K = {}, r = {}, q = {}, sigma = {}, T_minus_t = {}, M = {}, n = {}".format(S_t, K, r, q, sigma, T_minus_t, M, n))
	print("S_ave_t = {}, passing_time = {}, simulations = {}, repetitions = {}".format(S_ave_t, passing_time, simulations, repetitions))

	days = np.ceil((T_minus_t) * 365)  # in days
	days = int(days)
	dT = 1.0 / 365.0

	expected_asian_vals = []
	for rep in range(repetitions):
		print("#{} iteration.".format(rep))
		sim_results = []

		for i in range(simulations):
			S_beg = S_t
			sim_S_vals = []

			S_prev, S_next = S_beg, 0
			remained_days = days
			while (remained_days > 0):
				mean_lnS = np.log(S_prev) + (r - q - ( ( sigma ** 2.0 ) / 2.0) ) * dT # mean of lnST
				sigma_lnS = sigma * np.sqrt(dT)								   	    # sigma of lnST
				
				#if remained_days == days:
				#	print("mean = {}, sigmd = {}".format(mean_lnS, sigma_lnS))

				S_next = exp( random.gauss(mean_lnS, sigma_lnS) )

				sim_S_vals.append(S_next)
				S_prev = S_next
				remained_days -= 1
			
			#print("len of S_u_list = {}".format(len(S_u_list)))
			total_days = np.ceil( (passing_time + T_minus_t) * 365 ) + 1
			sim_S_ave = ( (passing_time * 365 + 1) * S_ave_t + sum(sim_S_vals) ) / (total_days)   # days = 0, 1, 2, ..., T * 365
			#print(max_S_u)
			payoff = max(sim_S_ave - K, 0)
			payoff = payoff * exp(- r * (T_minus_t) )
			sim_results.append(payoff)

		expected_asian = np.mean(sim_results)
		expected_asian_vals.append(expected_asian)

	sim_mean = np.mean(expected_asian_vals)
	sim_std  = np.std(expected_asian_vals)

	print("\nMonte Carlo results.")
	print("mean: {}, std: {}".format(sim_mean, sim_std))
	print("0.95 C.I. of a asian option: [{}, {}]".format(sim_mean - 2 * sim_std, sim_mean + 2 * sim_std))

			

File: hw5/hw5_monte.cpp
#include <iostream>
#include <fstream>
#include <math.h>
#include <cmath>
#include <stdio.h>
#include <algorithm>
#include <random>
#include <stdlib.h>
#include <time.h>

using namespace std;

void display(vector<double>& v) {
	int size = v.size();
	for (int i = 0; i < size; i += 1) {
		printf("%f,  ", v[i]);
	}
	printf("\n");
}

double mean_vector(vector<double>& v) {
	double ret = 0.0;
	int size = v.size();
	for (int i = 0; i < size; i += 1) {
		ret += v[i];
	}
	return ret / size;
}

double std_vector(vector<double>& v) {
	double ret = 0.0;
	int size = v.size();
	double mean = mean_vector(v);

	for (int i = 0; i < size; i += 1) {
		ret += pow(v[i] - mean, 2.0);
	}
	return sqrt(ret / (size));
}

double max_vec(vector<double>& v) {
	double max_val = -10e07;
	int size = v.size();
	for (int i = 0; i < size; i += 1) {
		double cur = v[i];
		if (cur > max_val)
			max_val = cur;
	}
	return max_val;
}

int main(int argc, char **argv) {
	double S_t, K, r, q, sigma, T_minus_t, S_ave_t, passing_time;
    int M, n, n_sim, n_rep;
	double u, d, p, dT;
    double passing_period;

    scanf("%lf %lf %lf %lf %lf %lf %d %d %lf %lf %d %d", &S_t, &K, &r, &q, &sigma, &T_minus_t, &M, &n, &S_ave_t, &passing_time, &n_sim, &n_rep);

	printf("Basic requirement: Monte Carlo simulation\n");
    printf("S_t          = %f\n", S_t);
    printf("K            = %f\n", K);
    printf("r            = %f\n", r);
    printf("q            = %f\n", q);
    printf("sigma        = %f\n", sigma);
    printf("T_minus_t    = %f\n", T_minus_t);
    printf("M            = %d\n", M);
    printf("n            = %d\n", n);
    printf("S_ave_t      = %f\n", S_ave_t);
    printf("passing_time = %f\n", passing_time);
    printf("n_sim        = %d\n", n_sim);
    printf("n_rep        = %d\n", n_rep);

	dT = (T_minus_t) / (double) n;
    passing_period = passing_time / (dT);
	u = exp(sigma * sqrt(dT));
	d = 1/u;
	p = (exp((r-q) * dT) - d)/(u - d);

	printf("u: %f, d: %f, p: %f, dT: %f\n", u, d, p, dT);

	default_random_engine generator;
	generator.seed(time(NULL)); // reset seed

	vector<double> expected_option_vals;
	for (int rep = 0; rep < n_rep; rep++) {
		//printf("#%d iteration.\n", rep);
		vector<double> sim_results;

		for (int i = 0; i < n_sim; i++) {
			vector<double> S_t_list;
            //S_t_list.push_back(S_t);
            
            double S_prev = S_t, S_next;
            
			int remained_days = n; // my way
			while (remained_days > 0) {
				//printf("dt = %f\n", dT);
				double mean_lnS_t = log(S_prev) + (r - q - ( pow(sigma, 2.0) / 2.0) ) * dT; // mean of lnST
				double sigma_lnS_t = sigma * sqrt(dT);								   	    // sigma of lnST

				normal_distribution<double> normal(mean_lnS_t, sigma_lnS_t);
				S_next = exp( normal(generator) );

				S_t_list.push_back(S_next);
				S_prev = S_next;
				--remained_days;
			}

            double sum = S_ave_t * passing_period + 
                         accumulate(S_t_list.begin(), S_t_list.end(), 0.0);

			//double S_ave = sum / (passing_period + n + 1); // S_t + (remained_days) of S_next => n + 1 values
			//double S_ave = sum / (passing_period + n + 1); // S_t + (remained_days) of S_next => n + 1 values
			double S_ave = sum / (passing_period + n ); // S_t + (remained_days) of S_next => n + 1 values

			double payoff = max(S_ave - K, 0.0);
			payoff = payoff * exp(-r * (T_minus_t) );
			sim_results.push_back(payoff);
		}

		//printf("len of sim_results = %d\n", sim_results.size());
		double expected_option = mean_vector(sim_results);
		expected_option_vals.push_back(expected_option);
	}

	double sim_mean = mean_vector(expected_option_vals);
	double sim_std  = std_vector(expected_option_vals);
   
    printf("-----------------------------------------\n");
    printf("### Answer for Monte Carlo Simulation ###\n");
	printf("mean: %f, std: %f\n", sim_mean, sim_std);
	printf("0.95 C.I. [%f, %f]\n", sim_mean - 2 * sim_std, sim_mean + 2 * sim_std);
    printf("-----------------------------------------\n");
	return 0;
}
